engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/package.scala

```diff
@@ -9,23 +9,62 @@ package object callcaching {
   case class HashValue(value: String)
   case class HashResult(hashKey: HashKey, hashValue: HashValue)
 
+  object HashValue {
+    implicit class StringMd5er(unhashedString: String) {
+      def md5HashValue: HashValue = {
+        val hashBytes = java.security.MessageDigest.getInstance("MD5").digest(unhashedString.getBytes)
+        HashValue(javax.xml.bind.DatatypeConverter.printHexBinary(hashBytes))
+      }
+    }
+  }
+
   private[callcaching] case class CacheResultMatchesForHashes(hashResults: Set[HashResult], cacheResultIds: Set[MetaInfoId])
   private[callcaching] case class CacheResultLookupFailure(reason: Throwable)
 
-  sealed trait CallCachingMode { def activity: Option[CallCachingActivity]; def readFromCache = false; def writeToCache = false; def lookupDockerHashes: Boolean = false }
+  sealed trait CallCachingMode {
+    def activity: Option[CallCachingActivity]
+
+    def readFromCache = false
+    def writeToCache = false
+    def hashDockerNames = false
+    def lookupDockerHashes: Boolean = false
+    def hashFilePaths: Boolean = false
+    def hashFileContents: Boolean = false
+
+    /**
+      * Return an equivalent of this call caching mode with READ disabled.
+      */
+    def withoutRead: CallCachingMode
+  }
+
   object CallCachingMode {
-    def apply(read: Boolean, write: Boolean, lookupDockerHashes: Boolean): CallCachingMode = (read, write) match {
+    def apply(read: Boolean, write: Boolean, hashDockerNames: Boolean, lookupDockerHashes: Boolean, hashFilePaths: Boolean, hashFileContents: Boolean): CallCachingMode = (read, write) match {
       case (false, false) => CallCachingOff
-      case (false, true) => WriteCache(lookupDockerHashes)
-      case (true, false) => ReadCache(lookupDockerHashes)
-      case (true, true) => ReadAndWriteCache(lookupDockerHashes)
+      case (false, true) => WriteCache(hashDockerNames, lookupDockerHashes, hashFilePaths, hashFileContents)
+      case (true, false) => ReadCache(hashDockerNames, lookupDockerHashes, hashFilePaths, hashFileContents)
+      case (true, true) => ReadAndWriteCache(hashDockerNames, lookupDockerHashes, hashFilePaths, hashFileContents)
     }
   }
 
-  case object CallCachingOff extends CallCachingMode { override def activity = None; }
+  case object CallCachingOff extends CallCachingMode {
+    override def activity = None
```

Comment:
I don't care enough to say to change it (feel free not to), but I'm chiming in here just to advocate what I think is a better pattern on ADTs. I find that for ADT hierarchies with simple overrides like this the FP style of implementation is cleaner, e.g.

    sealed trait parent {
      def foo = this match {
        case childA => "foo"
        case childB => "bar"
     }
    }
    case object childA extends parent
    case object childB extends parent

over the OO ADT style here.